### ISSUE TITLE
Specification for parse_kv_items and parse_kv_config

### ISSUE THREAD

**Issue by ypid** [debops/debops#1461]
`parse_kv_items` and `parse_kv_config` was introduced by @drybjed and is now a core part of DebOps. People working with DebOps need a clear understanding of it. Documentation what the filters do are contained in the actual role documentation but are specific to the role then. This will require people writing new roles to actually look into the code to figure out what it does and how to use it in their own roles. As already proposed by @tasosalvas, specification of the filters is needed defining exactly what their behavior is. I went for a hands-on approach when the filters where introduced and wrote the unit tests to reverse and "document" what they do. I would still be interested in a design document about our direction.

And maybe a "Not invented here" approach, but no better alternative comes to mind for me currently to solve all of the things that the filters do.

https://github.com/debops/debops/blob/master/ansible/roles/ansible_plugins/filter_plugins/debops_filter_plugins.py

I just want to track this here. Again, no pressure. I know maintaining a project is hard and this is a FOSS project. My idea is more that people interested in DebOps maybe can reverse the filters as well and not write unit tests this time but think about the problem and come up with a design document.

**Comment by drybjed** [comment]
Ah, the "magic sauce" of DebOps. :-) I'll try to write a rough description of why the filters were created and their generalized functionality. This will probably end up in the "Developer guide" section of the documentation since the contents of the `ansible_plugins` role will be moved to `ansible/plugins/` directory at some point to be accessible via Ansible Collection mechanisms.

@tasosalvas Did you write anything about the plugins yet? We can then combine the documentation together. If not that's fine as well.

**Comment by tasosalvas** [comment]
I have notes and a plan, I've been putting it off as work stuff had really piled up. I will try to get a first draft together. (just woke up)

**Comment by drybjed** [comment]
Good morning. :) OK then, I'll wait for your draft then. Let's see what you got. And don't worry, you don't need to do it right away if something else needs your attention.

**Comment by tasosalvas** [comment]
I fully agree with @ypid, digging into `parse_kv_config` really untied my hands as a debops user. The unit tests helped immensely.

I have a few paragraphs on:
* the input types the filter accepts
* the resulting mappings reserved attributes
* merging/weighing behavior
* probably some more stuff mixed with the `bind` role's notes

I haven't turned them into `.rst` yet. It's not too much work, I just have some trouble task switching when I get into something.
No need to duplicate this part, though, I got it.

Let's talk design, yeah.


## The goal
> How I learned to stop worrying and love `parse_kv_*`.

Understanding `parse_kv_` is useful for two distinct levels of working with the project, and I believe each of them needs its own doc:

* When purely configuring debops through inventory, it is a tool that allows the end user to do all sorts of magic without touching tasks or templates.
* When writing roles, apart from solving common issues in a DRY way, it allows granting flexibility to the end user in a consistent way they can git good at.

The canary, of course, is seeing a ton of role `defaults-detailed.rst` being simply reiterations of advanced `parse_kv_config` usage. We could totally gut the role docs if we consolidated them in one place.

Funnily, that is not where `parse_kv_` is most prominent. By the time users have to deal with one of those advanced lists, they've used it a ton of times in those sloppy-looking `*_combined_*` lists that live in every role.

The objective, as I see it, is to turn this universal configuration style into a first-class feature. A promise to the debops end user.


## User facing: Universal Configuration
> Way easier than regular expressions!

The user facing doc doesn't need to bother with `parse_kv_` details, but with usage. **Universal configuration** is not so much the filter, but its input.

Off the top of my head, this will include:
* An overview of the feature's utility: aggregate lists of unique items

A real life usage examples part:
* the `default` pattern: Modifying or removing default values
* the `all/group/host` pattern: Building up additive mappings
* using `.state` wherever you please
* using `.weight` wherever you please

And a reference part:
* Input styles (strings, kv items, arbitrary mappings)
* Fields in the input mappings that affect filter behavior
* A *see the dev doc for more* link. Or leave with something useful :)


## Dev facing: Beware, our users have superpowers
> Don't `| flatten` me bro!

I meean, if I see `combined` I have the right to feed it mappings just for lulz.

The developer side of the feature is mostly about keeping the promises made on the user-facing side.

Again, goal oriented and not trying to replace the unit tests.

* A *go read the user doc* link :p
* An overview of what an output mapping looks like
* A few examples of usage in templates and tasks
* conventions (such as `.list`, for lists not to be iterated)
* Where to use it and where not -- I say go mad, but what do I know ;)

Another little project I have in mind is the **Writing Roles** doc; A big part of that will be an exhaustive list of core roles you *must* know before writing one that deploys a fancy new toy.

I would suggest we work on this on the same document for now, but long-term I think it belongs in the *it's dangerous to go alone, take this* checklist.


## Benefits and implications
I would normally worry about using a bespoke custom tool to do all the things, but the filter is used almost everywhere already. Even if an Ansible update rendered parts of its usage obsolete, I bet we'd still have corner cases for which we'd want a lite version of it. So we may well embrace it. We also have tests!

Bringing it to the forefront would allow everyone to reason better about it, spot missing or excessive features, maybe enable usage patterns we haven't yet thought of -- The codebase itself certainly has plenty of examples of creative usage.

Not directly in the scope of this conversation, but the filter's name is a bit misleading, as its usage has probably diverged quite a bit from its original inception... Sometimes it's acutally `parse_k_`, sometimes it's `parse_kvvv_`.
Also `parse`, like, doesn't mean anything on a public function.
I think it could simply be called `| config` and `config_items`.


## Yes I write a lot
So, that's my idea. I was hoping to submit it together with a draft but didn't find the time and ended up sitting on it for quite a while... funny story, a major part of what was blocking me on the `bind` role is not being able to force myself to write another duplicate of the `kv` doc.

I think the best way to go about it is to have the discussion, then I lay down the draft based on what we decide (hopefully towards the end of this week), then you two can fact-check it and make additions without having to do much of the gruntwork.

**Comment by tasosalvas** [comment]
I also have a secret plan to make @drybjed like blockquotes ;)

**Comment by drybjed** [comment]
> The canary, of course, is seeing a ton of role `defaults-detailed.rst` being simply reiterations of advanced `parse_kv_config` usage. We could totally gut the role docs if we consolidated them in one place.

The problem I see with this is that on the "other side" of the `parse_kv_*` pipeline sits a program which has its own ideas about what each option does and how configuration files are processed. We can classify some of these programs by the format they are using (YAML, INI) but there are many exceptions to this. For example, LDAP configuration via `ldap_entry`/`ldap_attrs` tasks still needs its own explanation in the documentation. A "general-purpose" overview of how the filters work will probably be very helpful though, to familiarize the users with its concepts.

> Funnily, that is not where `parse_kv_` is most prominent. By the time users have to deal with one of those advanced lists, they've used it a ton of times in those sloppy-looking `*_combined_*` lists that live in every role.

One potential of `_combnined_` lists I see is integration with external configuration sources like `etcd` or `consul`. Just write a lookup template that gets the data from `etcd`, parses it and outputs configuration suitable for `parse_kv_items` filter - you just integrated a DebOps role with your `etcd` environment without touching the role at all.

> The objective, as I see it, is to turn this universal configuration style into a first-class feature. A promise to the debops end user.

I envisioned adding `parse_kv_*` filters to be included in Ansible itself when it is sufficiently polished and proven in DebOps. But now with Ansible going into the Collections route it will probably be sufficient to expose it via DebOps Collection - that way when people want to use it in their own setups they don't have to keep a separate copy of it, and we still have control over the filters and can fix any issues that are found.

> I would normally worry about using a bespoke custom tool to do all the things, but the filter is used almost everywhere already. Even if an Ansible update rendered parts of its usage obsolete, I bet we'd still have corner cases for which we'd want a lite version of it. So we may well embrace it. We also have tests!

The big push for the filter was the idea that roles and playbooks cannot be modified directly, only through the inventory. In combination with DebOps essentially "taking over" big configuration files to be able to template them effectively, and to avoid forcing users to copy large portions of role default variables to the inventory to just modify one or two values, I had to come up with something... Incidentally the filters were malleable enough to be useful in many different configuration patterns, and here we are.

I mostly look at these filters now as missing pieces of infrastructure between the `lineinfile` and `template` Ansible modules. Using them effectively requires a large invesment in copying over the configuration of an application and transforming it into the format acceptable by the filters, but in turn you gain the ability to configure that application with pin-point accuracy.

> Not directly in the scope of this conversation, but the filter's name is a bit misleading, as its usage has probably diverged quite a bit from its original inception... Sometimes it's acutally `parse_k_`, sometimes it's `parse_kvvv_`.
> Also `parse`, like, doesn't mean anything on a public function.
> I think it could simply be called `| config` and `config_items`.

I can tentatively agree with this. The name came from parsing a list of key-value entries into a dictionary which is then converted back into a list. The `parse_kv_config` filter came first to handle single configuration file (`/etc/postfix/main.cf`), and then `parse_kv_items` happened as an extension to be able to generate more complex configuration inside a different configuration file (`/etc/postfix/master.cf`) and it all went snowballing from there.

A generic `config` filter name could have been an issue in the past, since a script that modifies DebOps roles to be able to work as a Ansible Collection needs to find the filter names and change them from `parse_kv_items` to `debops.debops.parse_kv_items`. But when DebOps switches to a "Collection first" approach, that modification will not be needed anymore.

Actually, I think that we could add `config` and `config_items` as "aliases" [here](https://github.com/debops/debops/blob/master/ansible/roles/ansible_plugins/filter_plugins/debops_filter_plugins.py#L484-L488) and it should just work. That would let us avoid the large name change everywhere.

> So, that's my idea. I was hoping to submit it together with a draft but didn't find the time and ended up sitting on it for quite a while... funny story, a major part of what was blocking me on the `bind` role is not being able to force myself to write another duplicate of the `kv` doc.

I find myself in this exact predicament each time I write a new role. The code itself is ready to go in a day or two, but then documenting everything... Ugh. I'm painfully aware that the role needs to be documented to be usable, and if I don't do it and just switch to a different thing it will never get done, so I always try to force myself to write the docs but it sometimes takes a while. It's a very large roadblock from time to time. Any suggestions how to deal with it?

> I think the best way to go about it is to have the discussion, then I lay down the draft based on what we decide (hopefully towards the end of this week), then you two can fact-check it and make additions without having to do much of the gruntwork.

Sounds good to me. You have an excellent writing style, looking forward for your first draft. :)

**Comment by tasosalvas** [comment]
> The problem I see with this is that on the "other side" of the parse_kv_* pipeline sits a program which has its own ideas about what each option does and how configuration files are processed.

Yeah, the general purpose bit is the easiest to achieve at this point. I'm hoping that once we can take this for granted it will be easier to identify more complex patterns to pass on as style advice on the dev doc. Roles are always going to have *some* logic unique to them. As long as they all massage variables in the same way where possible, their own documentation will be the place to deal with the quirks of the program they're deploying.

Generally I don't think the filter should care about what the roles do with its output; Its emergent flexibility pretty much stems from that limited scope.

I was thinking we could have a standard way to signify vars that are aggregated in `parse_kv_` lists, as well as vars that are entry points (`_combined_`). Ideally two little icons linking to the user doc. A short sentence could do for now, and we can `sed` it out later I guess. Anyway, that's a step 2 kind of thing, once we have the docs we're talking about in place and can start cutting.


> integration with external configuration sources like `etcd` or `consul`

I haven't used either of those, so this is just a random thought, but I do think this layer of abstraction between inventory vars and role input would be just the place for switching the source. The extra hoops the role templates have to go through by always working with mappings might be just enough to allow them to not care where it came from.

A feature like this would be a reason to actually pass *everything* through the filter, which is not where we're at right now (I'm not saying I don't like it, though). Top level vars would need to be bundled somewhere (hopefully on the fly and not in inventory). Lists of non-unique things (like `include` statements) could need separate handling too; Currently roles that need those accept lists and take manual care to emit the config line multiple times, which might be incompatible with what other config tools supply.

Overall I think the current plan will improve consistency in role variable input but not really bring it to the level where roles can switch sources without getting bloated with more conditional logic. The `_dependent_` parts might be a good first target, I guess.


> The big push for the filter was the idea that roles and playbooks cannot be modified directly, only through the inventory. 

I haven't really used ansible without debops for anything but tests, but the whole workflow involving modifying roles was really annoying to me and a big reason why debops fit my own approach. I hadn't really made the connection that this is what we needed it for, but it's definitely worth mentioning in the user doc.


> I think that we could add config and config_items as aliases

I think we make `parse_kv` call `config` and emit a deprecation warning, `sed` the thing out of existence in the codebase, then deprecate the old name in the next major release. No real benefits in keeping around two ways of doing the same thing.


> I find myself in this exact predicament each time I write a new role. The code itself is ready to go in a day or two, but then documenting everything... [...] Any suggestions how to deal with it?

I usually have an RDD-inspired approach, it's the duplication that gets me. I take it as a sign that I should not touch anything more and reevaluate the whole design. I tend to overanalyze things. The `style-force` fixes were a way for me to get used to juggling multiple branches, and I think getting this `parse_kv_` doc out of the way will get me back into finishing `bind`.

So I have a feeling the tips I need could be the exact opposite to the ones you would need... which, as I see it, is a good thing, as it allows us to bring different strengths to the process.

Yours and ypid's interest on this one certainly helped get me going, so I think community definitely helps. :)

**Comment by tasosalvas** [comment]
Hey, I started writing and just realized **Configuration** is already a thing in the *user manual*...
I think that existing section would fit better next to **Installation** anyway.

I was also thinking of adding a classifier to it, like "project configuration". It does cover global and user config though... maybe "DebOps configuration"? I don't know. Maybe moving it next to installation is enough.

I'm open to alternate name suggestions for *Universal Configuration* too, though I do think it sounds pretty cool.

**Comment by drybjed** [comment]
I think that the exact section or the title can be decided later, it's not a problem to rearrange the index. :)

**Comment by tasosalvas** [comment]
Yeah, I'm not stuck, I just thought I'd state the issue and my current thinking on it.

I'm also having some trouble explaining:

* item.option
* item.options
* item.separator
* item.section

I can see some of those used in `postfix`, but I'm not sure why they belong in `parse_kv` and not the role's logic.
I would expect `section` to at least interact with `weight`, but glancing at the filter I don't think that's the case.

I'm taking today to work on the text, so the first draft will hopefully be ready tonight, then I can spend some time poking those mystery spots.
I bet there are things I'm missing, but it also feels like the filter might have some cruft that could be simplified without really affecting behavior.

**Comment by drybjed** [comment]
Well, let's see then...

The `item.option` parameter is useful in cases where configuration file can contain multiple options with the same name. In such cases you cannot use `item.name` multiple times directly because the entries will be merged. A solution for this is to name the entries with an inique `item.name` each and specify `item.option` as the "actual" option name. Then in the tempate it can be used as:
```yaml
{{ '{} = {}'.format((item.option | d(item.name)), item.value) }}
```
Then when you want to override a specific option you still can, after looking up the actual `item.name` of course.

The `item.options` parameter is used in the `parse_kv_items` filter as an "entry point" for `parse_kv_config` filter. It acts as the "second level configuration" and its entries are merged the same way the first level is. A common use case is with the `conf.d/` configuration style, where the first level defines the specific configuration files, and `item.options` is a list which defines the contents of the configuration files. You can actually enable more parameters to act like this if you want, to be able to do even more crazy things - see `rsnapshot` role for a good example of multi-level merges.

The `item.separator` parameter is just for aesthetics. Usually a boolean, can be used to denote that a given option needs a blank space or two in the configuration file, or there should be a commeted out line that separates sections... I like to generate pixel-perfect configuration files as much as possible, that way you can use diff to easily spot differences. Sometimes configuration files have a particular layout and `item.separator` helps keep that somewhat consistent.

The `item.section` parameter is also usually an aesthetics thing. Sections in configuration files are nothing new and this parameter helps with easily defining what section a given option should belong to. The `rsyslog` role has `/etc/rsyslog.conf` template with a great example of this - there's a separate list of sections which define the ordering in the configuration file. Then at the generation time, template goes over the section list and filters out only the entries that belong in a given section. The `weight` parameter in this case will only let you move the entries around inside a given section to keep the configuration file consistent - this is especially important if the order of the configuration options matters.

Don't forget that there are many more roles that use the filters. You can use the Changelog to see what recent roles have been added to DebOps, most likely they use the filters in one way or another.

I can definitely say that filters are overengineered a bit and you don't usually use all of their features at once. But there are some corner cases or unusual solutions where you can find them most useful.

**Comment by tasosalvas** [comment]
Oh, I see, weight doesn't need to weigh section items separately, as the template is going to act on one section at a time anyway.

Heh, I think I'll need some time for your explanations to sink in and figure out a good way and place to express them in the user doc, but fortunately I have a bunch of simpler examples to go through before I get to them.

Wouldn't some of these work the same way if the filter was unaware of them and they existed only as conventions/suggestions in the role-writing doc? My reasoning is that implementing roles need to be aware of all reserved output mapping keys so they don't spit them out in the resulting config, so keeping the list short might have some value.

If we're going to be deprecating the `parse_kv_` name, it might be a good opportunity to break some compatibility and actually do a gradual migration of roles to a leaner `config` filter written to the emergent spec.

Anyway, back to writing.

**Comment by drybjed** [comment]
> Oh, I see, weight doesn't need to weigh section items separately, as the template is going to act on one section at a time anyway.

You still might want to reorder the entries wthin the section.

> Heh, I think I'll need some time for your explanations to sink in and figure out a good way and place to express them in the user doc, but fortunately I have a bunch of simpler examples to go through before I get to them.

I told you that it will be hard to write a general-purpose documentation for these filters. :-)

> Wouldn't some of these work the same way if the filter was unaware of them and they existed only as conventions/suggestions in the role-writing doc? My reasoning is that implementing roles need to be aware of all reserved output mapping keys so they don't spit them out in the resulting config, so keeping the list short might have some value.

The problem is that each and every parameter and keyword in the filters was added for a specific purpose or because a quirk in a certain application configuration needed to be handled. Of course if you write configuration for a simple application that uses one configuration file based on key=value format you won't need all of these bells and whistles.

But there are certain applications that need all that complexity. The original filters were written for the `postfix` role's `/etc/postfix/main.cf` and `/etc/postfix/master.cf` templates, respectively. The latter one is very complicated and you still can modify every single parameter there with surgical precision, because the filters allow for it.

At the end of the day, filters get a simple list of parameters and output a structured data about configuration which needs to be written somewhere. How you use that data, in what way, it depends all on you - you can go with a template file, a set of tasks, whatever.

As for reserved mapping keys - think of it this way: using the filters you don't configure an application directly; you instead configure a tool which will configure that application. You need to look at it "one level up", in which case any keywords we reserve for filter operations will become irrelevant because you define the actual configuration options using the `name` and `value` dictionary keys, for example.

> If we're going to be deprecating the `parse_kv_` name, it might be a good opportunity to break some compatibility and actually do a gradual migration of roles to a leaner `config` filter written to the emergent spec.

If we start removing functionality from filters, things will start to break left and right. Complexity is like energy - it cannot be increased or decreased, only moved around. The filters are complex so that templates and tasks can be simple in return without sacrificing functionality.

**Comment by tasosalvas** [comment]
> Complexity is like energy - it cannot be increased or decreased, only moved around.

I have to disagree. Perhaps if we were omniscient and could cram the whole stack in our working memory that would be the case _[citation still needed]_. Being mere fleshy humans, _perceived_ complexity is all we really have access to and it certainly doesn't share this property. I refer you to [this ancient, sacred text](http://www.ariel.com.au/jokes/The_Evolution_of_a_Programmer.html).

But still, moving complexity around between roles and this filter that runs multiple times on every role can be a net benefit.


> You still might want to reorder the entries wthin the section.

Yes, that was my initial point; The filter reorders _all_ entries and the section ordering only emerges when the other sections are discarded from the resulting mixed list _by the role_. The filter doesn't separate the sections _then_ reorder, and doesn't need to. 

At this point the `section` doesn't really affect the filter's inner workings, and could as well be an arbitrary key the implementing role uses, knowing that the filter will leave it untouched. A simple `{% for item in list if item.section | d('unknown') == 'mysection' %}` in the role would then result in the same behavior.

> The problem is that each and every parameter and keyword in the filters was added for a specific purpose

Same goes for `separator`, where no calculations are performed apart from an `item.get('separator', false)`.

I understand `| d()` was not as simple to use in the olden days. Would you add another keyword today for a similar purpose?


> You need to look at it "one level up", in which case any keywords we reserve for filter operations will become irrelevant because you define the actual configuration options using the name and value dictionary keys, for example.

I look at it through the use-case that cares about it: List items as arbitrary mappings with multiple keys, where all unhandled keys are spit out as key-value pairs in config, reserved terms have to be handled or omitted, and clashing terms have to be worked around.

Plenty of applications (with `bind` being a good example, where your take on it also took a similar approach) have configurations so sprawling and arbitrarily nested that handling every possible valid option would be highly impractical in terms of maintenance.


> If we start removing functionality from filters, things will start to break left and right.

That's why I suggested we combine this with the name change and deprecation warning. The simplified filter could live alongside the old one for a while and roles can migrate to it on a case-by-case basis.

We could track its coverage through the deprecation warnings the old `parse_kv` will emit. We can begin with it stripped down to the keys that are actually used in internal calculations, then reintroduce complexity to it as needed. If we can achieve, say, 90% coverage without a feature, then it can be argued that the feature belongs to the implementing roles.

**Comment by drybjed** [comment]
> Yes, that was my initial point; The filter reorders _all_ entries and the section ordering only emerges when the other sections are discarded from the resulting mixed list _by the role_. The filter doesn't separate the sections _then_ reorder, and doesn't need to.
> 
> At this point the `section` doesn't really affect the filter's inner workings, and could as well be an arbitrary key the implementing role uses, knowing that the filter will leave it untouched. A simple `{% for item in list if item.section | d('unknown') == 'mysection' %}` in the role would then result in the same behavior.

Ah, I see what you mean. After checking in the filter, `section` and `separator` could definitely be dropped from the "distinct" status to a regular key and not much would have changes. Some roles would probably error out when an expected value isn't there, but that's an easy fix. Might be a good idea to drop these.

> > The problem is that each and every parameter and keyword in the filters was added for a specific purpose
> 
> Same goes for `separator`, where no calculations are performed apart from an `item.get('separator', false)`.
> 
> I understand `| d()` was not as simple to use in the olden days. Would you add another keyword today for a similar purpose?

Probably not. :) Well, the filters are overengineered - now that we know how they are used in practice, we can trim the fat a bit and make them easier to understand.

> That's why I suggested we combine this with the name change and deprecation warning. The simplified filter could live alongside the old one for a while and roles can migrate to it on a case-by-case basis.
> 
> We could track its coverage through the deprecation warnings the old `parse_kv` will emit. We can begin with it stripped down to the keys that are actually used in internal calculations, then reintroduce complexity to it as needed. If we can achieve, say, 90% coverage without a feature, then it can be argued that the feature belongs to the implementing roles.

Sounds good. I'm working on improved `debops` script and digging this Python thing more and more. I'm not sure how to handle deprecation warnings yet, I'll have to check that out in Ansible documentation. I guess some kind of [deprecation warning exception](https://docs.python.org/3/library/warnings.html) would be in order.

**Comment by tasosalvas** [comment]
Yeah, that's the one. I'm not sure where we would tell ansible to run python stuff with a `-Wa` flag, but I guess it will probably listen to the `PYTHONWARNINGS` environment var. Then we hope ansible devs use it too every once in a while, or we get flooded with deprecation warnings Python emits for their own code. :D

My suggested approach towards re-doing the filters would be (surprise) RDD/TDD. We write tests based on what the doc promises to the users, then filters to pass those tests.

My next issue has to do with [this test](https://github.com/debops/debops/blob/de2b3e57babf15be65c94bccd013ad4230e4da87/ansible/roles/ansible_plugins/filter_plugins/debops_filter_plugins.py#L530), as I was documenting the filter's recursion behavior.

Although the filter will process lists stored in `.value`, it does not merge them but simply gives the `parse_kv_` treatment to the last list provided.

Many occurrences I have noticed in the codebase simply avoid recursion altogether by placing lists in other variables, such as `.list`. I understand `.options` implements the behavior just fine, and the `merge_keys` argument allows `_items` to do it on other keys as well.

Is that wholesale overriding of `.value` a desired behavior I should document? Is it used somewhere? The way I see it it's a confusing middle way between `.options` and `.list`.

Anyway, I'm skipping it for now to get the things I actually understand down.

**Comment by drybjed** [comment]
> Yeah, that's the one. I'm not sure where we would tell ansible to run python stuff with a `-Wa` flag, but I guess it will probably listen to the `PYTHONWARNINGS` environment var. Then we hope ansible devs use it too every once in a while, or we get flooded with deprecation warnings Python emits for their own code. :D

Interesting. This can be a toggle in the DebOps configuration file, I think. We will see how new scripts work first.

> My suggested approach towards re-doing the filters would be (surprise) RDD/TDD. We write tests based on what the doc promises to the users, then filters to pass those tests.

Hmm, sure... RDD? Rubber Duck Debugging? I know what TDD means.

I still need to properly grasp how to test stuff in Python. I currently don't do tests while the new scripts are shaping up, I'll see what can be done about them later.

> My next issue has to do with [this test](https://github.com/debops/debops/blob/de2b3e57babf15be65c94bccd013ad4230e4da87/ansible/roles/ansible_plugins/filter_plugins/debops_filter_plugins.py#L530), as I was documenting the filter's recursion behavior.
> 
> Although the filter will process lists stored in `.value`, it does not merge them but simply gives the `parse_kv_` treatment to the last list provided.

This is done so that you can control the order of the items in the list via `weight` and `copy_id_from` parameters. This was needed in Postfix where you might want to order the lookup tables in a certain order, and to do that you need to be able to specify that this new item you are adding needs to be placed before that other item, and so on.

You can then extract the final list in the template with something like:
```yaml
{{ item.value | selectattr('state', 'equalto', 'present') | map(attribute='name') | list }}
```

> Many occurrences I have noticed in the codebase simply avoid recursion altogether by placing lists in other variables, such as `.list`.

Hmmm, any examples? There might be a reason for that particular choice, if you point me at a role I might reason about it further.

> I understand `.options` implements the behavior just fine, and the `merge_keys` argument allows `_items` to do it on other keys as well.
> 
> Is that wholesale overriding of `.value` a desired behavior I should document? Is it used somewhere? The way I see it it's a confusing middle way between `.options` and `.list`.

Yes, that behaviour for lists in the `item.value` parameters is desired and should be documented.

**Comment by tasosalvas** [comment]
[Readme Driven Development](https://tom.preston-werner.com/2010/08/23/readme-driven-development.html). I guess it can be considered a sloppy offshoot of Test-driven. It has probably stuck with me as I'm not too great at directing my attention, so having notes on what I was doing is great help.

I was mostly saying the next version's tests should be geared towards making sure the intended usage works, instead of being an attempt to reverse-engineer the filter's behavior after the fact. Not that they would end up being very different, but they would have the doc as a long-form explanation to back up what they're testing for.

In this specific instance @ypid's `unittest` setup is a fine starting point and part of standard Python. [pytest](https://docs.pytest.org/en/latest/) is the bigger guns.

Researching `bind` I noticed roles from galaxy implement a testing framework named [molecule](https://molecule.readthedocs.io/en/latest/). I meant to ask if we care about it but decided not to kick that hornet's nest at that moment. Long term it'd probably be a better fit for us than pure Python solutions. I haven't really researched it, though.

> Hmmm, any examples? There might be a reason for that particular choice, if you point me at a role I might reason about it further.

lol cancel that, I saw `name`, `list` pairs in your `bind` role, from before `parse_kv_` times. Then I misunderstood `.value`'s behavior and ended up keeping them as a bad habit in my own implementation. They're not everywhere, I just spent a ton of time with them -- Thanks for pointing it out. :)

**Comment by ypid** [comment]
> Long term it'd probably be a better fit for us than pure Python solutions. I haven't really researched it, though.

My go to for more complex unit testing is nosetests in Python.

**Comment by tasosalvas** [comment]
`nose` declares itself inactive and suggests `nose2` or `pytest`.
`nose2` declares itself immature and suggests also checking out `pytest`.

They're all quite functional, but if I was picking one atm I think I'd follow the suggestions.

**Comment by ypid** [comment]
Thanks, then I am not up-to-date. I will check when I next work with unit tests.

**Comment by drybjed** [comment]
> [Readme Driven Development](https://tom.preston-werner.com/2010/08/23/readme-driven-development.html). I guess it can be considered a sloppy offshoot of Test-driven. It has probably stuck with me as I'm not too great at directing my attention, so having notes on what I was doing is great help.

I dunno... The [first README of the initial project that became DebOps](https://github.com/debops/debops/tree/eb42149555). I'm not sure how specific it would have to have been to be ready for writing the actual code. But that methodology is probably better suited for smaller and more focused projects, like the filters themselves.

> I was mostly saying the next version's tests should be geared towards making sure the intended usage works, instead of being an attempt to reverse-engineer the filter's behavior after the fact. Not that they would end up being very different, but they would have the doc as a long-form explanation to back up what they're testing for.

Agreed, when proper documentation is written it will be easier to design the filters and their tests "to spec", so to speak.

> Researching `bind` I noticed roles from galaxy implement a testing framework named [molecule](https://molecule.readthedocs.io/en/latest/). I meant to ask if we care about it but decided not to kick that hornet's nest at that moment. Long term it'd probably be a better fit for us than pure Python solutions. I haven't really researched it, though.

I've looked at Molecule a bit and in some cases it's the same type of tool as the `jane` script currently used by DebOps. However where DebOps tests are "vertical" - linting is done globally as the first step and after that roles are tested via GitLab CI but only those that were modified, Molecule seems to be geared towards "horizontal" testing - each role is linted, syntax-checked and tested in a live environment separately one by one. With a large and growing number of roles, test times might become prohibitive - that's why putting some of the tests together as a first step (linting, syntax) might make more sense with DebOps.

> > Hmmm, any examples? There might be a reason for that particular choice, if you point me at a role I might reason about it further.
> 
> lol cancel that, I saw `name`, `list` pairs in your `bind` role, from before `parse_kv_` times. Then I misunderstood `.value`'s behavior and ended up keeping them as a bad habit in my own implementation. They're not everywhere, I just spent a ton of time with them -- Thanks for pointing it out. :)

Ah, the "before" times. :-) That was a wild period of experimentation, growth and learning.

**Comment by tasosalvas** [comment]
Hehe, yeah, I found it a good fit on design driven projects, such as games and websites for clients, where implementation is largely irrelevant to what the end user sees. On stuff presenting technical challenges or requiring experimentation I end up rewriting whole sections disappointed at how much I didn't know what I was talking about when I wrote them.

I hadn't really looked at `jane`. :) If you're doing all this with `bash`, I think you could port it to Python with a hand tied behind your back. I was mostly thinking of not-pure-python because I was considering setting up vagrant boxes as fixtures to be a big fuss.

I would argue it's `jane` that's horizontal and `molecule` that's vertical according to your description, in terms of, say the BFS/DFS algorithms, or the "vertical slice" concept; But meh, semantics.

I'm a bit burnt out to finish today, but our discussions have left me with a much better understanding of the feature and I've had some good progress. I'll hack at it a bit more and hopefully wrap it up tomorrow.

**Comment by tasosalvas** [comment]
`real_weight` always equals `id` in the outputted mappings.

I realize there are midpoints in the filter's calculations where this is not the case, but it would be cleaner if the key was namespaced as `_real_weight` when we reimplement, and perhaps even removed before being handed to the role.

`.weight` itself has some value for debug purposes in the output, I guess.

**Comment by tasosalvas** [comment]
[test_parse_kv_items_copy_id_from](https://github.com/debops/debops/blob/de2b3e57babf15be65c94bccd013ad4230e4da87/ansible/roles/ansible_plugins/filter_plugins/debops_filter_plugins.py#L852) isn't complex enough to test the feature.

**Comment by tasosalvas** [comment]
damn, ok, I misunderstood something in each of my two previous comments. Getting there :)

**Comment by tasosalvas** [comment]
Things got a bit silly. Am I proud of myself? Yes, yes I am :P

**Comment by tasosalvas** [comment]
Now that I don't feel like a slacker, let's fight :P

The test I linked above has the exact same output and passes if `copy_id_from` is omitted by both the input and result.
@ypid you wrote it, any insight into what you were testing for?

After thoroughly reviewing the `weight` implementation, I propose we rename the field to `item.jenga` until it passes this test:

```yaml
# obviously whoever declared all these explicit weights wants to override source order
- name: third
  weight: 30
- name: second
  weight: 20
- name: first
  weight: 10
```

```yaml
- id: 0
  name: third
  real_weight: 30
  separator: false
  state: present
  weight: 30
- id: 10
  name: second
  real_weight: 30
  separator: false
  state: present
  weight: 20
- id: 20
  name: first
  real_weight: 30
  separator: false
  state: present
  weight: 10
```

Do point me to any other tool that mixes up initial source ordering with weights in this way.
I'm very interested in how they rationalize this feature to the end user.

The issue gets compounded by the fact that these lists are meant to be aggregated and both the dev and user only control half of them at any given point.

`copy_id_from` comes as a patch to these inconsistencies but extensive usage, especially when done simultaneously by the role dev and the end user, is a recipe for off-by-**ten** errors. Weighing items in an `all/group/host` context is straight up playing jenga.

The only use cases I see for the feature as it stands are:
* Meticulously weigh everything in a list and never touch it again for fear you will break something
* Weigh items in increments of 1000 and hope you never get enough items to make even that number insignificant

I'm sorry if I'm mean.
I understand it got the job done, but if we're making this a user facing feature it has to be more intuitive and similar to other weight systems.

Anyway, here's the spec for the next iteration:
* All items have a default weight of 0 and preserve the source order of their first occurrence
* Items with an explicit negative weight float to the top of the list and sort themselves there
* Items with an explicit positive weight sink to the bottom of the list and sort themselves there
* An `.anchor_to` property makes items sort themselves right after (if their weight is positive) or right before (if their weight is negative) the element referenced by name.

If we really need an `id` to enumerate the items, it gets added after sorting is done.

The benefit is that the computer does the math and the human controlling it does not.
How's that for moving complexity around?

I might rewrite the filter myself on the spec next month if you're busy with other stuff.

**Comment by drybjed** [comment]
Oh, I like that algorithm better than the exiting one already. When can we add it to the filters? Roles will need to be reviewed to update them for new behaviour, but that's an easy step.

**Comment by tasosalvas** [comment]
We add it to the new `config` filters, we migrate the roles to it. I would initially consider anything that worked with the old one to be incompatible, although it may pleasantly surprise us, I guess.

I kinda want to do the rewrite, no offense but I think I can do prettier python than what's in the existing filters.
I need to rush my web presence and a CV in the next couple days, though, submissions end midnight Aug 23.

**Comment by drybjed** [comment]
@tasosalvas Of course, take care of yourself first. I can also do prettier Python after a few weeks writing it, the filters were written a few years ago. :)

**Comment by tasosalvas** [comment]
15 years :P

I'd also like to standardize and unify the filter behaviors and arguments to avoid confusion like the one in our value merging discussion

**Comment by drybjed** [comment]
Sure, as long as all roles keep the existing features we might unify that behaviour, namely lists being additive in both filters.

**Comment by tasosalvas** [comment]
But you said you do need non-merging behavior, no? Although merging gives us all the tools to clean up unwanted items by `state`, I guess.

**Comment by ypid** [comment]
> The test I linked above has the exact same output and passes if copy_id_from is omitted by both the input and result.
@ypid you wrote it, any insight into what you were testing for?

Git log tells me that at the time I wrote the tests I worked on #835.

> I kinda want to do the rewrite, no offense but I think I can do prettier python than what's in the existing filters.

:+1:
